### PR TITLE
feat(cmf): add saga router to handle side effects

### PR DESCRIPTION
**Pull request by talend-sromain**
**What is the problem this PR is trying to solve?**
To handle some side effects ( for example route change ) we don't have some elegant solution to catch actions with some context to make side effect behaviours

**What is the chosen solution to this problem?**
Add saga router to define in your application an association of routes & actions to yield. as it use generator. The generator would be canceled on route change. more documentation about redux saga here : https://redux-saga.js.org/

**Please check if the PR fulfills these requirements**
- [x] The PR commit message follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Related design / discussions / pages (not in jira), if any, are all linked or available in the PR

**[ ] This PR introduces a breaking change**

### CODE REVIEW

File: packages/cmf/package.json [hunk 1]
```diff
@@ -1,109 +1,106 @@
...
-    "redux-storage-decorator-filter": "^1.1.8",
-    "redux-storage-decorator-immutablejs": "^1.0.4",
-    "redux-storage-engine-localstorage": "^1.1.4",
-    "redux-thunk": "2.2.0"
-  },
-  "version": "0.107.0",
-  "dependencies": {
-    "hoist-non-react-statics": "^1.2.0"
-  }
+	"name": "react-cmf",
```
Comment (jmfrancois) [inline packages/cmf/package.json]: the style for json is 
```js
[*.json]
indent_style = space
indent_size = 2
```
from the editorconfig.
So please use 2 space here

File: packages/cmf/src/sagaRouter/router.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
+import 'babel-polyfill';
```
Comment (jmfrancois) [inline packages/cmf/src/sagaRouter/router.js]: this should not be done here

File: packages/cmf/src/sagaRouter/router.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+			const routeFragment = routeFragments[index];
+			const match = matchPath(currentLocation.pathname, { path: routeFragment });
+			const maybeSaga = sagas[routeFragment];
+			if (shouldStartSaga(maybeSaga, match)) {
+				sagas[routeFragment] = {
+					saga: yield spawn(routes[routeFragment], match.params),
+					match,
+				};
+			} else if (shouldCancelSaga(maybeSaga, match)) {
+				if (maybeSaga) {
```
Comment (jare-talend) [inline packages/cmf/src/sagaRouter/router.js]: useful? we have already a test on maybeSaga in shouldCancelSaga to enter in this block

File: packages/cmf/src/sagaRouter/router.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+					yield cancel(maybeSaga.saga);
+				}
+			} else if (shouldRestartSaga(maybeSaga, match)) {
+				if (maybeSaga) {
```
Comment (jare-talend) [inline packages/cmf/src/sagaRouter/router.js]: same thing

File: packages/cmf/src/sagaRouter/router.js (hunk 3 above)
Comment (talend-sromain) [reply packages/cmf/src/sagaRouter/router.js]: not that usefull, i fix that

File: packages/cmf/src/sagaRouter/router.js (hunk 4 above)
Comment (talend-sromain) [reply packages/cmf/src/sagaRouter/router.js]: not that usefull, i fix that

File: packages/cmf/src/sagaRouter/matchPath.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+	const [url, ...values] = match;
+	const isExact = pathname === url;
+
+	if (exact && !isExact) return null;
+
+	return {
+		path, // the path pattern used to match
+		url: path === '/' && url === '' ? '/' : url, // the matched portion of the URL
+		isExact, // whether or not we matched exactly
+		params: keys.reduce((memo, key, index) => {
```
Comment (jare-talend) [inline packages/cmf/src/sagaRouter/matchPath.js]: you can use this:

params: keys.reduce((memo, key, index) => ({
...memo,
[key.name] = values[index],
});

Comment (talend-sromain) [reply packages/cmf/src/sagaRouter/matchPath.js]: it's code from react-router v4, i don't think it's usefull to replace it. As it's now, we have this file, but when migration on v4 would be done, this file would be replace by the one from the dependency

Comment (jare-talend) [reply packages/cmf/src/sagaRouter/matchPath.js]: ok

**Pull request merged**